---
title: "c3co"
author: "Morgane Pierre-Jean, Julien Chiquet, Henrik Bengtsson and Pierre Neuvial"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: c3co.bib
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
---

## Introduction

The `c3co` package implements a constraint dictionary learning problem to recover cancer subclones from several DNA copy number profiles described in @pierre-jean:c3co. The c3co model may be seen as an extension of the FLLat method of @nowak2011fused, and of the e-FLLat method of @masecchia2013dictionary, with the following original features:

* **interpretable weights**: we model each profile as a *convex combination* of latent profiles, making the corresponding weights directly interpretable as proportions of latent features;
* **parent-specific copy numbers**: we leverage the allelic signals available from SNP array or sequencing data in order to explicitly integrate parent-specific copy numbers (Olshen *et al*, 2011) in the model. 
* **segment-level**: we model tumor clonality at the level of copy number segments (not individual loci), which is the level of information at which such events occur.

Notes: 

* Although the c3co model is designed to deal with allelic signals, we emphasize that it is also applicable to data where only total copy number estimates are available, such as array-CGH data or low-pass sequencing data.

* The FLLat method is implemented in the R package `FLLat`. The e-FLLat method is implemented in Python and is available from its authors.

## Model 

The figure below illustrates the model used in the `c3co` package. Two heterogeneous tumor samples (green and yellow circles) are composed of a collection of normal cells (gray discs) and two cancer subclones (red triangles and blue squares). One of the cancer subclones is present in both tumor samples. 

![](img/features.png)
![](img/features2.png)

The corresponding (noiseless) copy number profiles are displayed in the figure below. They are given by a linear combination of the latent profiles. This Figure is adapted from @nowak2011fused.

![](img/model.png)
![](img/model2.png)


## Using the package
```{r, message=FALSE,warning=FALSE}
stopifnot(packageVersion("acnr") >= '0.2.6')
library("c3co")
library("ggplot2")
library("mclust")
library("reshape")
set.seed(9)
```

### Creating a synthetic data set

We start by defining the characteristics of the subclone profiles: profile length, number of subclones, breakpoint positions, and copy number states:

```{r}
len <- 500*10
nbClones <- 3
bkps <- list(c(100,250)*10, c(150,400)*10,c(150,400)*10)
regions <-list(c("(0,1)", "(0,2)","(1,2)"), c("(1,1)", "(0,1)","(1,1)"), c("(0,2)", "(0,1)","(1,1)"))
```

Then, we load an annotated data set from the `acnr` package.  The `buildSubclones` function can then be used to generate the subclone profiles with the above characteristics, by resampling from the annotated data set.

```{r}
dataAnnotTP <- acnr::loadCnRegionData(dataSet="GSE13372_HCC1143", tumorFraction=1)
dataAnnotN <- acnr::loadCnRegionData(dataSet="GSE13372_HCC1143", tumorFraction=0)
datSubClone <- buildSubclones(len, dataAnnotTP, dataAnnotN, nbClones, bkps, regions)
```
 
Note that the same can be done using another data set from the `acnr` package, see

```{r}
acnr::listDataSets()
```

 
```{r, eval=FALSE}
dataAnnotTP <- acnr::loadCnRegionData(dataSet="GSE11976_CRL2324", tumorFraction=1)
dataAnnotN <- acnr::loadCnRegionData(dataSet="GSE11976_CRL2324", tumorFraction=0)
datSubClone2 <- buildSubclones(len, dataAnnotTP, dataAnnotN, nbClones, bkps, regions)
```

```{r, echo=FALSE}
cols <- c("#AAAAAA33", "#00000099", "#AAAAAA33")
cex <- 0.3
pch <- 19 
clim <- c(0, 4)
blim <- c(-0.1, 1.1)
plot(datSubClone[[1]]$ct, col=cols[factor(datSubClone[[1]]$genotype)], cex=cex, pch=pch, ylab="TCN", ylim=clim, main="PSCN profile of one simulated subclone")
plot(datSubClone[[1]]$baft, col=cols[factor(datSubClone[[1]]$genotype)], cex=cex, pch=pch, ylab="BAF", ylim=blim)
```

Once the subclones are created, we can generate a weight matrix $W$ in order to build mixtures.

```{r, warning=FALSE, cache=FALSE}
W = rSparseWeightMatrix(nb.samp=10, nb.arch=3, sparse.coeff = 0.8)
datList <- mixSubclones(subClones=datSubClone, W)
str(datList[[1]])
```

Note that `datList` is a list of data frames with the following required columns : `c1,c2,tcn,dh,genotype`

### Inference of the c3co model parameters

Then the c3co method can be applied to the mixture data set. Let us choose the same grid for $\lambda_1$ and $\lambda_2$ and a grid from 2 to 6 for the number of subclones.

```{r c3co, warning=FALSE, message=FALSE}
l1 <- seq(from = 10, to = 10^-3, length = 10)
nb.arch <- 2:6
parameters.grid <- list(lambda=l1, nb.arch=nb.arch) 
res <- c3co(datList, parameters.grid, verbose=FALSE)
resC <- c3co(datList, parameters.grid, stat="TCN", verbose=FALSE)
```

For each $p$,  `c3co` only retains the combination of the penalization coefficients $(\lambda_1, \lambda_2)$ which minimizes the Bayesian Information Criterion (BIC) of the model. The next step is to choose the best $p$ (number of subclones). Following @nowak2011fused, we compare the models for different values of $p$ through their percentage of variance explained (PVE), and select the last $p$ before the final plateau of the PVE. In this example, it seems that the best is $\hat{p}=4$ (which is the true number of subclones).

```{r}
pvePlot(res, ylim=c(0.0,1))
```

```{r}
pvePlot(resC, ylim=c(0.0,1))
```

We can compare the true and the estimated weight matrices. We can easily recover a classification close to the truth with the inferred weight matrix. 

```{r heatmap_C1C2, fig.width=7, fig.height=5}
best <- 3
res.clustTRUE = hclust(dist(cbind(W, 1-rowSums(as.matrix(W)))),method="ward.D")
col = grDevices::colorRampPalette(RColorBrewer::brewer.pal(9, 'GnBu'))(100)
Wplot(res, idxBest = best, cexCol = 0.4)
```

```{r heatmap_TCN, fig.width=7, fig.height=5}
Wplot(resC, idxBest = best, cexCol = 0.4)
```

```{r heatmap_TRUE, fig.width=7, fig.height=5}
heatmap.3(cbind(W, 1-rowSums(as.matrix(W))), dendrogram="row", main="TRUE", Rowv=as.dendrogram(res.clustTRUE), col=col,scale="none")
```

If we look at the subclones in the dimension of parental copy numbers, we can recover the simulated alterations.

```{r,Z1Z2hat, fig.width=7.5, fig.height=5}
df <- createZdf(res, chromosomes=1, idxBest = best)
Zplot(df)
```

## Further analysis
### Performance  of clustering

```{r,clust}
WC1C2 <- res@fit[[best]]@W
WTCN <- resC@fit[[best]]@W
Wtrue <- cbind(W, 1-rowSums(as.matrix(W)))
adjustedRandIndex(cutree(hclust(dist(WC1C2), method="ward.D2"),4), cutree(hclust(dist(Wtrue), method="ward.D2"),4))
adjustedRandIndex(cutree(hclust(dist(WTCN), method="ward.D2"),4), cutree(hclust(dist(Wtrue), method="ward.D2"),4))
```
## Performance of alteration detection
```{r, fig.width=7.5, fig.height=5}
getAlteredSegments <- function(tol, zz1, zz2, c1Mean, c2Mean){
  loss <- ((c1Mean-zz1) >= tol)| ((c2Mean-zz2) >= tol)
  gain <- ((zz2 - c2Mean) >= tol)|((zz1-c1Mean) >= tol)
  ## coding gains as 3
  gain <- apply(gain, 2, function (gg) {
    idx <- which(gg)
    gg[idx] <- 3
    return(gg)
  })
  ## coding loh segments as 4 (gain + loss)
  pos <- loss+gain
  return(pos)
}

dataBest <- res@fit[[best]]
meanMin <- mean(df$CopyNumber[which(df$arch=="d"&df$stat=="Minor")])
getAlt <- getAlteredSegments(tol=0.2, dataBest@S$Z1, dataBest@S$Z2, 1, 1.1)
colnames(getAlt) <- sprintf("Subclone %s", letters[1:ncol(dataBest@W)])
df.Alt <- melt(getAlt)
df.Alt$x.min <- rep(res@bkp[[1]][-7], 4)
df.Alt$x.max <- rep(res@bkp[[1]][-1], 4)
ggplot(df.Alt, aes(xmin = x.min, xmax = x.max, ymin = 1, ymax =2, fill = factor(value))) + geom_rect() + facet_grid(X2~.)+theme(axis.text.y = element_blank(), axis.ticks = element_blank())+scale_fill_discrete(labels= c("Normal", "Loss", "Gain", "LoH"), name="States")
```

## Session information

```{r}
sessionInfo()
```

## References


